File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/BoundingRectanglePaintable.java
package edu.uci.ics.jung.visualization.layout;

import com.google.common.graph.Network;
import edu.uci.ics.jung.layout.model.LayoutModel;
import edu.uci.ics.jung.visualization.MultiLayerTransformer;
import edu.uci.ics.jung.visualization.RenderContext;
import edu.uci.ics.jung.visualization.VisualizationModel;
import edu.uci.ics.jung.visualization.VisualizationServer;
import edu.uci.ics.jung.visualization.util.ChangeEventSupport;
import java.awt.*;
import java.awt.geom.Rectangle2D;
import java.util.List;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

public class BoundingRectanglePaintable<N> implements VisualizationServer.Paintable {

  protected RenderContext rc;
  protected Network graph;
  protected LayoutModel<N> layoutModel;
  protected List<Rectangle2D> rectangles;

  public BoundingRectanglePaintable(RenderContext rc, VisualizationModel<N, ?> visualizationModel) {
    super();
    this.rc = rc;
    this.layoutModel = visualizationModel.getLayoutModel();
    this.graph = visualizationModel.getNetwork();
    final BoundingRectangleCollector.Nodes<N> brc =
        new BoundingRectangleCollector.Nodes<>(rc, visualizationModel);
    this.rectangles = brc.getRectangles();
    if (layoutModel instanceof ChangeEventSupport) {
      ((ChangeEventSupport) layoutModel)
          .addChangeListener(
              new ChangeListener() {

                public void stateChanged(ChangeEvent e) {
                  brc.compute();
                  rectangles = brc.getRectangles();
                }
              });
    }
  }

  public void paint(Graphics g) {
    Graphics2D g2d = (Graphics2D) g;
    g.setColor(Color.cyan);

    for (Rectangle2D r : rectangles) {
      g2d.draw(rc.getMultiLayerTransformer().transform(MultiLayerTransformer.Layer.LAYOUT, r));
    }
  }

  public boolean useTransform() {
    return true;
  }
}

File: jung-visualization/src/main/kotlin/edu/uci/ics/jung/visualization/layout/BoundingRectanglePaintable.kt
package edu.uci.ics.jung.visualization.layout

import edu.uci.ics.jung.layout.model.LayoutModel
import edu.uci.ics.jung.visualization.MultiLayerTransformer
import edu.uci.ics.jung.visualization.RenderContext
import edu.uci.ics.jung.visualization.VisualizationModel
import edu.uci.ics.jung.visualization.VisualizationServer
import edu.uci.ics.jung.visualization.util.ChangeEventSupport
import java.awt.Color
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.geom.Rectangle2D
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener

open class BoundingRectanglePaintable<N : Any>(
    private val rc: RenderContext<*, *>,
    visualizationModel: VisualizationModel<N, *>
) : VisualizationServer.Paintable {

    private val layoutModel: LayoutModel<N> = visualizationModel.getLayoutModel()
    private var rectangles: List<Rectangle2D>

    init {
        val brc = BoundingRectangleCollector.Nodes<N>(rc, visualizationModel)
        this.rectangles = brc.getRectangles()
        if (layoutModel is ChangeEventSupport) {
            (layoutModel as ChangeEventSupport).addChangeListener(object : ChangeListener {
                override fun stateChanged(e: ChangeEvent) {
                    brc.compute()
                    rectangles = brc.getRectangles()
                }
            })
        }
    }

    override fun paint(g: Graphics) {
        val g2d = g as Graphics2D
        g.setColor(Color.cyan)

        for (r in rectangles) {
            g2d.draw(rc.getMultiLayerTransformer().transform(MultiLayerTransformer.Layer.LAYOUT, r))
        }
    }

    override fun useTransform(): Boolean = true
}
